src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs

```diff
@@ -289,6 +304,255 @@ public struct S
+public struct S
+{
+    public int i;
+
+    // error
+    public readonly override string ToString() => (i++).ToString();
+    public readonly override int GetHashCode() => (i++).GetHashCode();
+    public readonly override bool Equals(object o) => (i++).Equals(o);
+}
+";
+            var comp = CreateCompilation(csharp);
+            comp.VerifyDiagnostics(
+                // (7,52): error CS1604: Cannot assign to 'this' because it is read-only
+                //     public readonly override string ToString() => (i++).ToString();
+                Diagnostic(ErrorCode.ERR_AssgReadonlyLocal, "i").WithArguments("this").WithLocation(7, 52),
+                // (8,52): error CS1604: Cannot assign to 'this' because it is read-only
+                //     public readonly override int GetHashCode() => (i++).GetHashCode();
+                Diagnostic(ErrorCode.ERR_AssgReadonlyLocal, "i").WithArguments("this").WithLocation(8, 52),
+                // (9,56): error CS1604: Cannot assign to 'this' because it is read-only
+                //     public readonly override bool Equals(object o) => (i++).Equals(o);
+                Diagnostic(ErrorCode.ERR_AssgReadonlyLocal, "i").WithArguments("this").WithLocation(9, 56));
+        }
+
+        [Fact]
+        public void ReadOnlyMethod_GetPinnableReference()
+        {
+            var csharp = @"
+public unsafe struct S1
+{
+    static int i = 0;
+    ref int GetPinnableReference() => ref i;
+    void M1()
+    {
+        fixed (int *i = this) {} // ok
+    }
+    readonly void M2()
+    {
+        fixed (int *i = this) {} // warn
+    }
+}
+
+public unsafe struct S2
+{
+    static int i = 0;
+    readonly ref int GetPinnableReference() => ref i;
+    void M1()
+    {
+        fixed (int *i = this) {} // ok
+    }
+    readonly void M2()
+    {
+        fixed (int *i = this) {} // ok
+    }
+}
+";
+            var comp = CreateCompilation(csharp, options: TestOptions.UnsafeReleaseDll);
+            comp.VerifyDiagnostics(
+                // (12,25): warning CS8655: Call to non-readonly member 'GetPinnableReference' from a 'readonly' member results in an implicit copy of 'this'.
+                //         fixed (int *i = this) {} // warn
+                Diagnostic(ErrorCode.WRN_ImplicitCopyInReadOnlyMember, "this").WithArguments("GetPinnableReference", "this").WithLocation(12, 25));
+        }
+
+        [Fact]
+        public void ReadOnlyMethod_Iterator()
+        {
+            var csharp = @"
+using System.Collections.Generic;
+
+public struct S
+{
+    public int i;
+    public readonly IEnumerable<int> M1()
+    {
+        yield return i;
+        yield return i+1;
+    }
+
+    public readonly IEnumerable<int> M2()
+    {
+        yield return i;
+        i++;
+        yield return i;
+    }
+}
+";
+            var comp = CreateCompilation(csharp);
+            comp.VerifyDiagnostics(
+                // (16,9): error CS1604: Cannot assign to 'this' because it is read-only
```

Comment:
I really don't like this error message: I'm not trying to assign to `this`. I'm trying to assign to `i`, which is a member of `this`.

Reply:
You're right that this message should be better. Will mark it prototype so it can be addressed in a future PR.

Compare to when a field is marked `readonly`: https://sharplab.io/#v2:D4AQDABCCMDcCwAoEBmCAnApgQwCYHsA7AGwE8IBnAF3QFcBjKiAZSQG8kIuMcCTyAloSYCEiblAAsEALIAKAJSduHcRO4CA1JrESAvkj1A=